drgn_tools/kernfs_memcg.py

```diff
@@ -0,0 +1,224 @@
+# Copyright (c) 2024, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Kernfs_memcg
+--------------
+
+The ``drgn.helpers.linux.kernfs_memcg`` module provides helpers for working with the
+Linux memcg subsystem.
+"""
+import argparse
+from typing import Iterator
+
+from drgn import cast
+from drgn import container_of
+from drgn import FaultError
+from drgn import Object
+from drgn import Program
+from drgn.helpers.common.format import decode_enum_type_flags
+from drgn.helpers.linux import cgroup_path
+from drgn.helpers.linux import css_for_each_descendant_pre
+from drgn.helpers.linux import find_slab_cache
+from drgn.helpers.linux import for_each_page
+from drgn.helpers.linux import inode_path
+from drgn.helpers.linux import kernfs_path
+from drgn.helpers.linux import slab_cache_for_each_allocated_object
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.dentry import dentry_path_any_mount
+
+# cgroup subsystem id for memory cgroup, from kernel/cgroup/cgroup.c
+_MEMORY_CGRP_ID = 4
+
+
+def decode_css_flags(css: Object) -> str:
+    """
+    Get a human-readable representation of cgroup_subsys_state.flags
+
+    :param css: ``struct cgroup_subsys_state *``
+    """
+    CSS_DYING = css.prog_["CSS_DYING"]
+    flags = css.flags.value_()
+    if not flags:
+        # There is no dedicated flag value to indicate a zombie cgroup.
+        # A css.flags value of 0 indicates that cgroup destruction is
+        # complete but cgroup object has not been fully freed because
+        # of being pinned by some other object
+        return "ZOMBIE"
+
+    return decode_enum_type_flags(flags, CSS_DYING.type_, False)
+
+
+def for_each_kernfs_node(prog: Program) -> Iterator[Object]:
+    """
+    Iterate over all kernfs_node objects in the system.
+
+    :returns: Iterator of ``struct kernfs_node *`` objects.
+    """
+    kernfs_node_cache = find_slab_cache(prog, "kernfs_node_cache")
+    for kn in slab_cache_for_each_allocated_object(
+        kernfs_node_cache, "struct kernfs_node"
+    ):
```

Comment:
I have verified that rb_node within kernfs_node can't be used to iterate through all kernfs_node objects , even it can't be used to iterate through all kernfs_nodes under a single root. The reason being that kernfs_node has one rb_node member to link it in its paren't rbtree , but it also has a rb_root, kernfs_node.dir.children to put all its immediate children there i.e a child subtree is not hosted at kernfs_node.rb.rb_right or rb_left . This means that rb tree traversal stops at immediate child level and does not go deeper to iterate through children of immediate child. 
So we have to use the current method but ensure that we don't use it if slab caches are merged, which is not the case with UEKs  at least.

Reply:
Got it, that makes sense. I guess I'm still a bit unclear on why you need kernfs nodes. You have other helpers that iterate over all the memory cgroups themselves. Is the expectation that there may be some memory cgroups that appear as kernfs_nodes but aren't attached to the main cgroup hierarchy?

Reply:
> Got it, that makes sense. I guess I'm still a bit unclear on why you need kernfs nodes. You have other helpers that iterate over all the memory cgroups themselves. Is the expectation that there may be some memory cgroups that appear as kernfs_nodes but aren't attached to the main cgroup hierarchy?

Yes, the zombie memcgroups will not appear in memcgroup hierarchy, that is why we need check kernfs_node(s) to find zombies.

Reply:
Ok, that makes sense!